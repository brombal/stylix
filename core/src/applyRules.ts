import { StylixContext } from './StylixProvider';

export function flattenRules(ctx: StylixContext): string[] {
  return Object.values(ctx.rules).flatMap((val) => val.rules);
}

/**
 * Applies rules from given StylixContext to the <style> element.
 */
export default function applyRules(ctx: StylixContext): void {
  if (ctx.styleCollector) {
    const flattenedRules = flattenRules(ctx);
    ctx.styleCollector.length = 0;
    ctx.styleCollector.push(...flattenedRules);
    return;
  }

  if (ctx.ssr) return;

  const supportsAdoptedStylesheets = 'adoptedStyleSheets' in document;

  // If there's no style element, and we're in dev mode or legacy browser, create one
  if (!ctx.styleElement && (ctx.devMode || !supportsAdoptedStylesheets)) {
    ctx.styleElement = document.createElement('style');
    ctx.styleElement.className = 'stylix';
    if (ctx.id) ctx.styleElement.id = 'stylix-' + ctx.id;
    document.head.appendChild(ctx.styleElement);
  }

  // If there's a style element, use its stylesheet
  if (ctx.styleElement) ctx.stylesheet = ctx.styleElement.sheet as CSSStyleSheet;

  // No stylesheet yet, create one
  if (!ctx.stylesheet) {
    ctx.stylesheet = new CSSStyleSheet();
    document.adoptedStyleSheets.push(ctx.stylesheet);
  }

  if (ctx.devMode && ctx.styleElement) {
    const flattenedRules = flattenRules(ctx);
    ctx.styleElement.innerHTML = flattenedRules.join('\n');
  } else {
    const stylesheet = ctx.stylesheet;
    if (!stylesheet) return;
    const flattenedRules = flattenRules(ctx);
    if (stylesheet.replaceSync) {
      try {
        stylesheet.replaceSync(flattenedRules.join('\n'));
      } catch (e) {
        // Errors are ignored, this just means that a browser doesn't support a certain CSS feature.
        console.warn(e);
      }
    } else {
      // Legacy method
      while (stylesheet.rules?.length || stylesheet.cssRules?.length) {
        stylesheet.deleteRule(0);
      }
      for (const i in flattenedRules)
        try {
          stylesheet.insertRule(flattenedRules[i], +i);
        } catch (e) {
          // Errors are ignored, this just means that a browser doesn't support a certain CSS feature.
          console.warn(e);
        }
    }
  }
}
