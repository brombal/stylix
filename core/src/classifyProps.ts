import { useStylixContext } from './StylixProvider';

export function classifyProps(props: any, knownProps: Record<string, string>): [any, any] {
  const styles = {} as any;
  const other = {} as any;

  for (const prop in props) {
    // If prop is not a valid JSX prop, it must be a CSS selector
    if (!isValidJSXProp(prop) || (isStyleProp(prop, knownProps) && isStyleValue(props[prop]))) {
      styles[prop] = props[prop];
    } else {
      other[prop] = props[prop];
    }
  }

  return [styles, other];
}

export function useClassifyProps(props: any) {
  const ctx = useStylixContext();
  const [styles, other] = classifyProps(props, ctx.styleProps);
  return [styles, other];
}

/**
 * Determines if `value` is a recognized CSS property (can be standard CSS or custom Stylix prop).
 */
export function isStyleProp(prop: string, knownProps: Record<string, string>): boolean {
  return isValidJSXProp(prop) && simplifyStylePropName(prop) in knownProps;
}

export function isValidJSXProp(value: string): boolean {
  // Not an exact check, but mostly rules out complex css selectors
  return /^[a-z$][a-z0-9_-]*$/i.test(value);
}

export function simplifyStylePropName(value: string) {
  return value.toLowerCase().replace(/[^a-z]/gi, '');
}

/**
 * Tries to determine if `value` is likely to be a valid CSS property value.
 * We can't be 100% sure, but this should catch most cases.
 * There is a check here to make sure React elements do not pass the test, as this
 * has turned out to be a common case where a property like 'content' means something
 * to a component, but is also a valid CSS property.
 */
export function isStyleValue(value: any): boolean {
  return (
    typeof value === 'function' ||
    typeof value === 'string' ||
    typeof value === 'number' ||
    typeof value === 'boolean' ||
    typeof value === 'undefined' ||
    Array.isArray(value) ||
    // Check for plain objects, and make sure it doesn't have the $$typeof property (react elements are never valid as style values)
    (typeof value === 'object' && value.constructor === Object && !('$$typeof' in value))
  );
}
