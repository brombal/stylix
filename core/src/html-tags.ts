export default [
  'a',
  'abbr',
  'address',
  'area',
  'article',
  'aside',
  'audio',
  'b',
  'bdi',
  'bdo',
  'blockquote',
  'body',
  'br',
  'button',
  'canvas',
  'caption',
  'cite',
  'code',
  'col',
  'colgroup',
  'data',
  'dd',
  'del',
  'details',
  'dfn',
  'dialog',
  'div',
  'dl',
  'dt',
  'em',
  'embed',
  'fieldset',
  'figcaption',
  'figure',
  'footer',
  'form',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'header',
  'hgroup',
  'hr',
  'html',
  'i',
  'iframe',
  'img',
  'input',
  'ins',
  'kbd',
  'label',
  'legend',
  'li',
  'main',
  'map',
  'mark',
  'menu',
  'menuitem',
  'meter',
  'nav',
  'noscript',
  'object',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'picture',
  'pre',
  'progress',
  'q',
  'rt',
  'ruby',
  's',
  'samp',
  'section',
  'select',
  'slot',
  'small',
  'source',
  'span',
  'strong',
  'sub',
  'summary',
  'sup',
  'svg',
  'table',
  'tbody',
  'td',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'tr',
  'track',
  'u',
  'ul',
  'var',
  'video',
] as const;
