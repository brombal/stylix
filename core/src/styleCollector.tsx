import React, { createContext } from 'react';

export interface StyleCollector {
  collect: (element: React.ReactElement) => React.ReactElement;
  render: React.FC<React.ComponentProps<'style'>>;
  styles: string[];
}

export const styleCollectorContext = createContext<string[]>(null);

export function createStyleCollector() {
  const styles: string[] = [];
  const collector: StyleCollector = {
    collect: (element) => (
      <styleCollectorContext.Provider value={styles}>{element}</styleCollectorContext.Provider>
    ),
    render: (props: React.ComponentProps<'style'>) => (
      <style
        type="text/css"
        key={props.id || 'stylix'}
        {...props}
        dangerouslySetInnerHTML={{ __html: collector.styles.join(' ') }}
      />
    ),
    styles,
  };
  collector.render.displayName = 'StylixStyleCollectorRenderer';
  return collector;
}
