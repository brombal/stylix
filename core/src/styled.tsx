import React from 'react';

import { _Stylix } from './Stylix';
import {
  Extends,
  HtmlOrComponent,
  HtmlOrComponentProps,
  MappedProperties,
  StylixComponentMeta,
  StylixProps,
} from './types';

/**
 * A component that accepts Stylix props in addition to TProps (Stylix props override TProps).
 */
export type StylixStyledComponentWithProps<TProps> = React.FC<Extends<TProps, StylixProps>> &
  StylixComponentMeta;

/**
 * A component that accepts Stylix props in addition to the props of TComponent (Stylix props override TComponent props).
 */
export type StylixStyledComponent<TComponent extends HtmlOrComponent> =
  StylixStyledComponentWithProps<HtmlOrComponentProps<TComponent>>;

/**
 * Gets a union of all the props of TComponent that conflict with css style properties.
 */
export type ConflictingStyleProps<TComponent extends HtmlOrComponent> = keyof StylixProps &
  keyof HtmlOrComponentProps<TComponent>;

/**
 * A map of props of TComponent that conflict with css style properties to the name of the prop that should be passed to the original component.
 */
type PropMap<TComponent extends HtmlOrComponent> = {
  [key: string]: ConflictingStyleProps<TComponent>;
};

/**
 * A 'styled' component collects all style props and reduces them to a className which it passes as a prop to the
 * component. The new component accepts all styled props in addition to the original props of the component.
 * By default, all style props are consumed as styles, so if the original component accepts a prop that conflicts with a style prop, it will not be received by the component.
 * If you want a style prop to be passed directly to the original component (and maintain TS types from the original component), you
 * can specify it in the propMap.
 * must specify them in the config. You can also specify props to map from one name to another, so that the
 * resulting styled component can accept a non-style-prop name that is renamed and passed to the original component
 * as the prop that might otherwise conflict with a style prop.
 */
export function styled<
  TComponent extends HtmlOrComponent,
  TPropMap extends PropMap<TComponent> | undefined,
>(
  component: TComponent,
  /**
   * A style object to apply to the component.
   */
  styleProps?: StylixProps,
  /**
   * A map of new props to accept and the associated prop name to pass to the original component.
   */
  propMap?: TPropMap,
): React.FC<
  // Props include the props of TComponent, overridden by all style props, overridden by the values of propMap mapped to the types of their corresponding keys
  Extends<
    HtmlOrComponentProps<TComponent>,
    StylixProps,
    TPropMap extends PropMap<TComponent>
      ? MappedProperties<HtmlOrComponentProps<TComponent>, TPropMap>
      : object
  >
> &
  StylixComponentMeta {
  const Element: React.ElementType =
    typeof component === 'string' // HTML element
      ? (component as any)
      : typeof component === 'function' && !('$$typeof' in component) && component.length === 2 // React function component with ref parameter
        ? React.forwardRef(component)
        : component; // React function component
  const r: any = React.forwardRef((props: Record<string, any>, ref: React.Ref<unknown>) => {
    props = { ...props }; // copy props so we can delete props that are consumed by Stylix
    const $elProps: any = {};
    if (propMap) {
      for (const incomingProp in propMap) {
        const outgoingProp = propMap[incomingProp];
        $elProps[outgoingProp] = props[incomingProp];
        delete props[incomingProp];
      }
    }
    return _Stylix(
      {
        $el: <Element {...$elProps} />,
        ...props,
      },
      ref as any,
    );
  });
  r.displayName = `$.${
    (component as React.FC).displayName ||
    (component as React.FC).name ||
    (component as string).toString?.() ||
    'Unnamed'
  }`;
  r.__isStylix = true;
  return r;
}
