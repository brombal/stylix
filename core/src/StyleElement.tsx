import React from 'react';

import { flattenRules } from './applyRules';
import { useStylixContext } from './StylixProvider';

export function StyleElement(
  props: { styles: string[] } & Partial<React.HTMLProps<HTMLStyleElement>>,
) {
  const { styles, ...other } = props;
  return (
    <style type="text/css" {...other} dangerouslySetInnerHTML={{ __html: styles.join('\n') }} />
  );
}

export function RenderServerStyles(props: Partial<React.HTMLProps<HTMLStyleElement>>) {
  const ctx = useStylixContext();
  return <StyleElement styles={ctx.ssr ? flattenRules(ctx!) : []} {...props} />;
}
