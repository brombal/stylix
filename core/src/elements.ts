import htmlTags from './html-tags';
import { styled } from './styled';
import Stylix from './Stylix';

for (const i in htmlTags) {
  // Types are specified in ./types.ts, so we don't care what they type of htmlTags[i] is.
  // JSX.IntrinsicElements is a union of all HTML tags, so it is too complex for TypeScript to infer.
  const tag = htmlTags[i] as any;
  (Stylix as any)[tag] = styled(tag, undefined, {
    htmlContent: 'content',
    htmlTranslate: 'translate',
  } as any);
}
