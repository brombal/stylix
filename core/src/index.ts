export {
  type StylixProps,
  type StylixPropsExtensions,
  type Stylix$Component,
  type Stylix$elProp,
  type Stylix$elPropOptional,
  type Extends,
  type MappedProperties,
  type HtmlOrComponentProps,
} from './types';
export { useStylixContext, StylixProvider } from './StylixProvider';
export { StyleElement, RenderServerStyles } from './StyleElement';
import { StylixPublicContext } from './StylixProvider';
export type StylixContext = StylixPublicContext;
export { useStyles, useKeyframes, useGlobalStyles } from './useStyles';
export {
  defaultPlugins,
  customProps,
  type StylixPlugin,
  type StylixPluginFunctionContext,
} from './plugins';
export { mapObjectRecursive } from './util/mapObjectRecursive';
export { walkRecursive } from './util/walkRecursive';
export { createStyleCollector, styleCollectorContext, type StyleCollector } from './styleCollector';
export { classifyProps, useClassifyProps } from './classifyProps';
export { styled, type StylixStyledComponentWithProps, type StylixStyledComponent } from './styled';
import './elements';
export { default } from './Stylix';
