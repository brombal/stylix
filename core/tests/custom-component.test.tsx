import React from 'react';

import $, { StylixProps } from '../src';
import { renderStylix } from './util';

describe('Custom component', () => {
  it('should work with basic styles', () => {
    function MyComponent(props: StylixProps) {
      return <$.div {...props} />;
    }

    // Accepts style prop
    void (<MyComponent color="red" />);
    void (<MyComponent marginTop={5} color={['red']} />);
    void (<MyComponent margin-top={5} color={() => 'red'} />);

    // @ts-expect-error - color is wrong type
    void (<MyComponent color={0} />);
    // @ts-expect-error - color is wrong type
    void (<MyComponent color={[0]} />);
    // @ts-expect-error - color is wrong type
    void (<MyComponent color={() => 0} />);

    const [json, styles] = renderStylix(<MyComponent color="red" />);
    expect(json).toMatchSnapshot();
    expect(styles).toMatchSnapshot();
  });

  it('should work with extended styles', () => {
    function MyComponent(props: StylixProps<{ otherProp: string; color: 'onlyThis' }>) {
      return <$.div {...props} />;
    }

    // Accepts style prop
    void (<MyComponent otherProp="test" color="onlyThis" />);

    // @ts-expect-error - color is wrong type
    void (<MyComponent otherProp="test" color="red" />);
    // @ts-expect-error - otherProp is missing
    void (<MyComponent color="red" />);
    // @ts-expect-error - otherProp has wrong type
    void (<MyComponent otherProp={5} color="red" />);

    const [json, styles] = renderStylix(<MyComponent otherProp="foo" color="onlyThis" />);
    expect(json).toMatchSnapshot();
    expect(styles).toMatchSnapshot();
  });

  it('should map props to new name', () => {
    function MyComponent(
      props: StylixProps<{ color: 'must be this value'; otherProp: string }, { newColor: 'color' }>,
    ) {
      return <$.div {...props} />;
    }

    void (<MyComponent color="must be this value" otherProp="" newColor="color" />);

    // @ts-expect-error - color is wrong value
    void (<MyComponent color="red" otherProp="" />);
    // @ts-expect-error - otherProp & newColor missing
    void (<MyComponent color="must be this value" />);

    const [json, styles] = renderStylix(
      <MyComponent otherProp="foo" color="must be this value" newColor="color" />,
    );
    expect(json).toMatchSnapshot();
    expect(styles).toMatchSnapshot();
  });

  it('should map optional props', () => {
    function MyComponent(
      props: StylixProps<
        { color: 'must be this value'; otherProp: string },
        { newColor?: 'color' }
      >,
    ) {
      return <$.div {...props} />;
    }

    void (<MyComponent color="must be this value" otherProp="" newColor="color" />);
    void (<MyComponent color="must be this value" otherProp="" />);
  });
});
