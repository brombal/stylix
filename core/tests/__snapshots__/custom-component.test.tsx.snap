// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom component should map props to new name 1`] = `
<div
  className="stylix-1lntnqv"
  newColor="color"
  otherProp="foo"
/>
`;

exports[`Custom component should map props to new name 2`] = `
[
  ".stylix-1lntnqv {
  color: must be this value; }",
]
`;

exports[`Custom component should work with basic styles 1`] = `
<div
  className="stylix-ik0txt"
/>
`;

exports[`Custom component should work with basic styles 2`] = `
[
  ".stylix-ik0txt {
  color: red; }",
]
`;

exports[`Custom component should work with extended styles 1`] = `
<div
  className="stylix-10d44kw"
  otherProp="foo"
/>
`;

exports[`Custom component should work with extended styles 2`] = `
[
  ".stylix-10d44kw {
  color: onlyThis; }",
]
`;
